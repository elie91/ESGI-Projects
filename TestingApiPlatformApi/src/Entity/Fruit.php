<?php

namespace App\Entity;

use ApiPlatform\Core\Annotation\ApiResource;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Bridge\Doctrine\Validator\Constraints\UniqueEntity;
use Symfony\Component\Serializer\Annotation\Groups;
use Symfony\Component\Validator\Constraints as Assert;

/**
 * This is a dummy entity. Remove it!
 * @UniqueEntity(fields={"name"}, message="fruit name need to be unique")
 * @ApiResource(
 *     attributes={
 *          "normalizationContext"={"groups"={"fruit_read", "cart_read", "cart_fruit_read"}},
 *          "denormalizationContext"={"groups"={"fruit_write"}}
 *     }
 * )
 * @ORM\Entity
 *
 */
class Fruit
{
    /**
     * @var int The entity Id
     *
     * @ORM\Id
     * @ORM\GeneratedValue
     * @ORM\Column(type="integer")
     * @Groups({"cart_read"})
     */
    private $id;

    /**
     * @var string
     *
     * @ORM\Column(type="string", length=50)
     * @Assert\Length(
     *      min = 2,
     *      max = 50,
     *      minMessage = "Le nom du fruit doit faire au minimum {{ limit }} caractères",
     *      maxMessage = "Le nom du fruit doit faire au maximum {{ limit }} caractères",
     *      allowEmptyString = false
     * )
     * @Groups({"cart_read"})
     */
    public $name;

    /**
     * @ORM\OneToMany(targetEntity=CartFruit::class, mappedBy="fruit", cascade={"remove"})
     */
    private $cartFruits;

    public function __construct()
    {
        $this->cartFruits = new ArrayCollection();
    }

    /**
     * @return int
     */
    public function getId(): int
    {
        return $this->id;
    }

    /**
     * @param int $id
     */
    public function setId(int $id)
    {
        $this->id = $id;
    }

    /**
     * @return string
     */
    public function getName(): string
    {
        return $this->name;
    }

    /**
     * @param string $name
     */
    public function setName(string $name)
    {
        $this->name = $name;
    }

    /**
     * @return Collection|CartFruit[]
     */
    public function getCartFruits(): Collection
    {
        return $this->cartFruits;
    }

    public function addCartFruit(CartFruit $cartFruit): self
    {
        if (!$this->cartFruits->contains($cartFruit)) {
            $this->cartFruits[] = $cartFruit;
            $cartFruit->setFruit($this);
        }

        return $this;
    }

    public function removeCartFruit(CartFruit $cartFruit): self
    {
        if ($this->cartFruits->contains($cartFruit)) {
            $this->cartFruits->removeElement($cartFruit);
            // set the owning side to null (unless already changed)
            if ($cartFruit->getFruit() === $this) {
                $cartFruit->setFruit(null);
            }
        }

        return $this;
    }
}
