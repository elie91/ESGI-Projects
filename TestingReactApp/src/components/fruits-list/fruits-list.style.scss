.products-list {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(20rem, 1fr));
    grid-gap: 20px;
}
