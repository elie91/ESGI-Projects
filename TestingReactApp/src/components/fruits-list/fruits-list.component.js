import React from "react";
import {withRouter} from "react-router-dom";
import Fruit from "../fruit/fruit.component";
import {ENTRYPOINT} from "../../config/entrypoint";
import Cart from "../cart/cart.component";

export const FruitsList = ({fruits, carts, history, setFruits, setCarts, cartItems, addToCart}) => {

  const handleDelete = (event, fruit) => {
    fetch(ENTRYPOINT + fruit['@id'], {
      method: 'DELETE',
      headers: {
        'content-type': 'application/json'
      }
    })
      .then((result) => {
        if (result.ok) {
          const filtered = fruits.filter(fr => fr.id !== fruit.id);
          setFruits(filtered)
        }
      })
      .then(() => history.push('/'))
  }

  const handleDeleteCart = (event, cart) => {
    fetch(ENTRYPOINT + cart['@id'], {
      method: 'DELETE',
      headers: {
        'content-type': 'application/json'
      }
    })
      .then((result) => {
        if (result.ok) {
          const filtered = carts.filter(fr => fr.id !== cart.id);
          setCarts(filtered)
        }
      })
      .then(() => history.push('/'))
  }

  return (
    <>
      <h1 className="mt-3">Mes fruits</h1>
      <div className="row">
        {fruits.map(fruit => <Fruit key={fruit.id} fruit={fruit} addToCart={addToCart}
                                                         cartItems={cartItems}
                                                         handleDelete={handleDelete}/>)}
        {fruits.length === 0 && <div className="col-md-12"><p>Il n'y a pas de fruits</p></div>}
      </div>

      <h2 className="my-3">Mes paniers fruitÃ©s</h2>
      <div className="row">

        {carts.map(cart => <Cart key={cart.id} cart={cart} handleDelete={handleDeleteCart}/>)}

        {carts.length === 0 && <div className="col-md-12"><p>Il n'y a pas de panier de fruit</p></div>}

      </div>
    </>

  );

}
export default withRouter(FruitsList);
