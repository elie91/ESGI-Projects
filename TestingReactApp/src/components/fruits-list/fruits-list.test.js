import React from 'react';
import {shallow, mount} from 'enzyme';
import {FruitsList} from "./fruits-list.component";
import Fruit from "../fruit/fruit.component";
import {BrowserRouter} from 'react-router-dom';

describe('FruitList component', () => {

  let wrapper;
  let mockFruits;
  let mockAddToCart;
  let mockHistory;
  let mockProps;

  beforeEach(() => {
    mockFruits = [
      {
        '@id': "/fruits/14",
        '@type': "Fruit",
        cartFruits: [],
        id: 14,
        name: "test"
      },
      {
        '@id': "/fruits/15",
        '@type': "Fruit",
        cartFruits: [],
        id: 15,
        name: "toto"
      }
    ]
    mockAddToCart = jest.fn();
    mockHistory = {
      push: jest.fn()
    };

    mockProps = {
      fruits: mockFruits,
      carts: [],
      history: mockHistory,
      title: 'hats',
      cartItems: [],
      addToCart: mockAddToCart
    };

    wrapper = mount(
      <BrowserRouter>
        <FruitsList {...mockProps} />
      </BrowserRouter>
    );
  });

  afterEach(() => {
    wrapper.unmount();
  })

  it('should render FruitList component', () => {
    const wrapper = shallow(<FruitsList fruits={[]} carts={[]} cartItems={[]}/>);
    expect(wrapper).toMatchSnapshot();
  });

  it('sould render Fruit components with title', () => {
    expect(wrapper.find('.mt-3').text()).toBe('Mes fruits')
  });

  it('sould render Fruit components with fruits childs', () => {
    expect(wrapper.find(Fruit).length).toBe(mockFruits.length);
  });


});

