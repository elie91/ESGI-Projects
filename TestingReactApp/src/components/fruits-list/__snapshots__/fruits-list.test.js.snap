// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FruitList component should render FruitList component 1`] = `ShallowWrapper {}`;
